### PR TITLE
[Release] v3.0.0-rc.2

### PR DESCRIPTION
**Pull request by avtolstoy**
Review only

### CODE REVIEW

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 1]
```diff
@@ -1346,22 +1373,24 @@ int QuectelNcpClient::enterDataMode() {
...
         const char breakCmd[] = "+++";
         muxerDataStream_->write(breakCmd, sizeof(breakCmd) - 1);
         skipAll(muxerDataStream_.get(), 1000);
 
-        // Initialize AT parser
-        auto parserConf = AtParserConfig()
-                .stream(muxerDataStream_.get())
-                .commandTerminator(AtCommandTerminator::CRLF);
-        dataParser_.destroy();
-        CHECK(dataParser_.init(std::move(parserConf)));
```
Comment (sergeuz) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1362]: I'd reset the parser here.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 2]
```diff
@@ -1371,12 +1400,50 @@ int QuectelNcpClient::enterDataMode() {
...
-            // We've already switched into data mode
-            ok = true;
+            // IMPORTANT: ATO does not work all the time and we resume into
+            // some broken state where the modem does not reply to our PPP ConfReqs.
+            // What's worse is that ATH doesn't seem to work either, so we cannot
+            // disconnect an ongoing PPP session. Toggling DTR doesn't work either.
+            // As a workaround send a PPP LCP Echo Request, wait until we get a reply
+            // or at least something similar to it.
+            // If we don't get a reply, consider we are in a broken state. To get
+            // out of it we are going to close and re-open the data muxer channel.
```
Comment (sergeuz) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1414]: :grimacing:

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp [hunk 3]
```diff
@@ -1371,12 +1400,50 @@ int QuectelNcpClient::enterDataMode() {
...
+                        if (buf[i] == lcpEchoRequest[0]) {
+                            flagsSeen++;
+                        }
+                    }
+                    if (flagsSeen >= 2) {
+                        ok = true;
+                        break;
+                    }
+                } else if (r != SYSTEM_ERROR_TIMEOUT) {
+                    return r;
```
Comment (sergeuz) [inline hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1436]: So, just to clarify, we don't want to retry the echo request if we haven't received a reply for it within 1s?

File: hal/network/lwip/cellular/pppncpnetif.cpp [hunk 4]
```diff
@@ -322,6 +322,14 @@ void PppNcpNetif::pppEventHandler(uint64_t ev) {
...
         // Reset
         connectStart_ = 0;
+        const auto ncpMtu = celMan_->ncpClient()->getMtu();
+        if (ncpMtu > 0) {
+            const auto newMtu = std::min<unsigned>(mtu, ncpMtu);
+            if (mtu != newMtu) {
+                LOG(TRACE, "Updating MTU to: %u", newMtu);
+                client_.getIf()->mtu = newMtu;
+            }
+        }
```
Comment (sergeuz) [inline hal/network/lwip/cellular/pppncpnetif.cpp:332]: Nit: This condition can be rewritten in a slightly more readable way:
```cpp
if (ncpMtu > 0 && ncpMtu < mtu) {
    LOG(TRACE, "Updating MTU to: %u", ncpMtu);
    client_.getIf()->mtu = ncpMtu;
}
```

**Review by sergeuz — APPROVED** [review]

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp (hunk 1 above)
Comment (avtolstoy) [reply hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1362]: It should be in a cleanly initialized state here, right? Or more like now in https://github.com/particle-iot/device-os/pull/2300/files/c1cd1ed590d079711d469117684f0e0577195bc0#diff-186ce5679947e07e5ce6c3d4fce9cdce5ebd61e4c315bcf0658db6106ea54299R1383

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp (hunk 3 above)
Comment (avtolstoy) [reply hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1436]: Correct. Or don't receive something similar to a PPP frame: 0x7e ...... 0x7e

Comment (sergeuz) [reply hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1436]: Ah, I misread it as `== SYSTEM_ERROR_TIMEOUT`, sorry.

File: hal/network/ncp_client/quectel/quectel_ncp_client.cpp (hunk 1 above)
Comment (sergeuz) [reply hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1362]: Yeah, but we then reuse the same parser instance in a loop and ignore errors until the maximum number of attempts is reached. I don't think calling `dataParser_.reset()` would necessarily make a difference here, I just noticed that we usually call it in similar cases in other parts of the code.

Comment (avtolstoy) [reply hal/network/ncp_client/quectel/quectel_ncp_client.cpp:1362]: Added `dataParser_.reset()`

File: communication/src/messages.cpp [hunk 5]
```diff
@@ -316,7 +316,8 @@ size_t Messages::separate_response_with_payload(unsigned char *buf, uint16_t mes
 }
 
 size_t Messages::event(uint8_t buf[], uint16_t message_id, const char *event_name,
-             const char *data, int ttl, EventType::Enum event_type, bool confirmable)
+             const char *data, int ttl, EventType::Enum event_type, bool confirmable,
+			 size_t max_transmit_message_size)
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply communication/src/messages.cpp]: Moved into `Publisher`.

**Review by sergeuz — APPROVED** [review]

**Review by keeramis — APPROVED** [review]
Tested cellular part of the review 👍

**Pull request merged**